AMR indiv and fungal large files uploads management

anagperal opened the pull request:
### :pushpin: References

BASED ON CODE FROM https://github.com/EyeSeeTea/glass-dev/pull/363 by @MatiasArriola 

-   **Issue:** Closes #8699wradu [AMR indiv and fungal large files uploads/deletions management](https://app.clickup.com/t/8699wradu) 2025
-  **Issue:** Closes #869du2ypa [AMR indiv and fungal large files uploads/deletions management](https://app.clickup.com/t/869du2ypa) 2026

### :memo: Implementation
- install papaparse
- read CVS in chunks
- allow import tracker entities in sync mode
- in server uploads for AMR indiv and AMR fungal: use chunks to manage RIS files and save tracker entitites in sync mode
- Use concurrency when importing in async uploads
- Add skipSideEffects true in async uploads
Async upload workflow — one continuous sequence
- Add Stop-on-import-error 


 ### :repeat: Async Uploads Workflow (AMR – Individual / AMR – Fungal)

This document describes, step by step, how the server-side async upload script imports a
primary CSV file into DHIS2. The script processes uploads that were queued in the Datastore,
validating the whole file before importing and sending the data to the server in chunks.

The entry point is the `uploadDatasets` routine in `src/scripts/cliAsyncUploads.ts`.

#### One continuous sequence

##### Processing the queue

1. The script takes the pending uploads that still have retry attempts left and processes them
   **one at a time** — there is no parallelism between uploads; all chunking and concurrency
   happens inside a single upload.
2. For each upload it pauses about half a second, then re-checks the upload is still marked
   pending. If not, it is skipped.
3. If still pending, it loads the upload's record, determines whether it is a primary or
   secondary file, and finds its module (a missing module is an error).

##### Preparing the file

4. The upload is marked "uploading" and its CSV is downloaded from storage.
5. The configured **upload chunk size** is read — how many rows go to the server per request.
   It comes from the module's Datastore config (default 100; currently testing **300 vs 500** comming from Datastore config).
6. The program's configuration and validation rules are fetched once, up front.

##### Validation pass (whole file, before anything is imported)

7. The file is read in **large blocks of 5,000 rows**. Each block is fully validated — custom
   checks plus program-rule checks.
8. This pass is fail-fast: the instant a blocking error appears, reading stops, an error report
   is saved, and **the import is skipped entirely** — guaranteeing the whole file is validated
   before any data goes in.

##### Import pass (only if validation passed completely)

9. The file is read again in the same **5,000-row blocks**.
10. Each block is split into smaller chunks of the configured **upload chunk size (300/500)**.
11. Those chunks are sent to the server in **synchronous mode**, with the server's **rule engine
    and side effects turned off** (the rules were already checked in the validation pass),
    **6 at a time**. The concurrency is *batched*: six are sent, the script waits for all six to
    finish, then sends the next six.
12. **Stop-on-import-error:** if any chunk fails, the process stops immediately — no further
    chunks or blocks are imported — keeping whatever succeeded so far.
13. As it goes, it accumulates the results and the IDs of the created records. At the end, if any
    records were imported, their ID list is saved to a file and linked to the upload, and the
    import summaries are saved.

##### Finishing the upload

14. The accumulated results are inspected for blocking errors and whether any rows were actually
    imported.
15. The upload's final status is set accordingly — validated (clean), imported (partial, with
    errors), or uploaded (nothing imported) — and the upload is removed from the queue.
16. If anything failed during the whole sequence, the upload's attempt counter is increased
    instead; once it reaches the maximum (3 by default), it is removed from the queue and marked
    failed, otherwise it stays for a later run to retry.

> **Secondary files** follow the same shape, except AMR – Individual does a trial run first and
> only imports for real if the trial finds no blocking errors.

#### The knobs, in one place

| Knob | Value | Notes |
| --- | --- | --- |
| Uploads | one at a time | No concurrency between uploads. |
| File blocks | 5,000 rows (fixed) | Used for both the validation and import passes. |
| Upload chunk size | 300 / 500 (default 100) | Datastore config; rows per server request. |
| Concurrency | 6 chunks, batched | Currently **hardcoded**  but can be config by Datastore |

### :video_camera: Screenshots/Screen capture

### :fire: Testing

[transform_JPN_INDIV_2023.sh](https://github.com/user-attachments/files/24195311/transform_JPN_INDIV_2023.sh)

anagperal committed:
Install paparse

anagperal committed:
Allow sync tracker imports and validate and import AMR indiv and Fungal RIS CSV files in chunks

anagperal committed:
Change chunk sizes and add more logs about time

anagperal committed:
Remove time logs

anagperal committed:
Fix default date format in SAMPLE_DATE and the number of line when chunking

anagperal committed:
Add skipSideEffects true in async uploads

anagperal committed:
Import TEIs with concurrency instead of sequential for async uploads

anagperal committed:
Add more logs when future error

anagperal committed:
Get only program metadata once

anagperal committed:
Create a Map instead of use find

anagperal committed:
Get programRules metadata at the beginning for runProgramRuleValidations

anagperal committed:
Improve code format

anagperal committed:
Add support to node 18 for async upload script for server execution

ifoche commented:
Task linked: [CU-8699wradu BLOCKING AMR-I RIS upload error](https://app.clickup.com/t/8699wradu)

anagperal committed:
Add Stop-on-import-error in async uploads

anagperal committed:
fix(async-uploads): bound CSV chunk buffer and propagate import errors

The parser buffer drained only one 5000-row batch per PapaParse callback while
each callback delivered more rows than that, so the backlog grew unbounded and
crashed large files with an out-of-memory error before importing anything.
Also stop swallowing synchronous errors in the async uploads script, which let
a failed upload silently report success.

anagperal committed:
perf(async-uploads): build program rule static context once per chunk

Program rule validation rebuilt the same metadata-derived rule structures
(rules, variables, data elements, TEI attributes, enrollments index) on every
single event, which dominated validation time on large files. Add an
async-upload-only path that builds that context once per chunk of 5000 rows
and reuses it, cutting validation time per chunk by roughly 20x. The existing
per-event path used by AMC, the generic bulk-load importer and the synchronous
AMR import is untouched; getEffects now delegates to the same context-based
implementation the async path uses, so both share one validation codepath.

anagperal committed:
perf(async-uploads): rolling concurrency for tracker imports

Tracker chunk imports were dispatched in fixed batches that waited for the
slowest request in each batch before starting the next one, leaving other
concurrency slots idle. Add a rolling-pool variant of parallel accumulation
that keeps the configured number of requests in flight continuously, and wire
the async AMR import to it. Concurrency is now read from the module's
datastore config (asyncUploadMaxConcurrency) instead of being hardcoded,
so it can be tuned per instance without a deploy.

anagperal committed:
fix(async-uploads): treat asyncUploadMaxConcurrency 0 as unset

fluture's parallel() rejects a concurrency of 0 with a synchronous exception,
so an accidental 0 in the module's datastore config crashed the async import
outright instead of falling back to the default, unlike the sibling
uploadChunkSize fallback just above it which already used the same guard.

anagperal committed:
Merge development branch and solve conflicts

MatiasArriola reviewed (changes requested):
Thanks @anagperal! Nice work on this, I appreciate the effort you've put into supporting large file uploads.

Here are a few items worth reviewing before we merge:

### 1. Handling non-CSV files in `RISIndividualFungalDataCSVDefaultRepository.ts`

`getFromBlobInChunks` fails in case of non-csv files. Previously xlsx files were supported through `getFromBlob` (which is not used anymore). The previous method wouldn't work with big xlsx files, but maybe we want to still support xlsx files. Also xlsx files still accepted at UI level.

Maybe we can modify `getFromBlobInChunks` to fallback to `getFromBlob` (even though we don't stream xlsx files):

```typescript
if (isCsvFile(blob)) {
  // existing path
} else {
  return this.getFromBlob(dataColumns, blob).flatMap(rows => {
    // process rows in chunkSize slices, awaiting onChunk for each
  });
}
```


### 2. Stream abortion logic in `CSVUtils.ts`
CSV import errors do not reliably halt processing. When onChunk rejects, parser.abort() is invoked, but PapaParse's abort() still triggers the complete callback. This causes remaining buffered rows to drain and trigger onChunk again.

The flow right now:

1. parseCsvBlobInChunks() buffers rows in currentChunk.
2. If onChunk() fails or returns false, parser.abort() is called.
3. PapaParse triggers complete.
4. The complete hook calls drainBufferedChunks(true), processing whatever is left in currentChunk.

Suggested Fix:

- Explicitly track a terminal state (e.g., isAborted, hasError).
- In the complete callback, skip flushing buffered rows if parsing was aborted or if onChunk() requested a stop.
- Only flush in complete during a normal, successful execution path.

### 3. Row number resetting in `usecases/data-entry/amr-individual-fungal/common.ts`
Date-validation error line numbers reset for every CSV chunk. fileLineStart is ignored in the date-error branch, so errors after the first 5,000 rows report incorrect lines.

Suggested fix: 
```typescript
const line = fileLineStart + index;
return validateAllDateFieldsInRow(row, AMR_INDIVIDUAL_FUNGAL_DATE_COLUMNS, line).map(err => ({
    error: err.message,
    line: err.row,
}));
```

anagperal committed:
fix(async-uploads): support xlsx chunking, reliable CSV abort, cross-chunk line numbers

- getFromBlobInChunks: fall back to full in-memory read for non-CSV (xlsx)
  files and emit rows in chunkSize slices instead of erroring
- parseCsvBlobInChunks: track terminal state so PapaParse's post-abort
  complete callback no longer replays buffered rows through onChunk after a
  stop or error
- runCustomValidations: offset date-format error line numbers by
  fileLineStart so they stay correct beyond the first chunk
